services/management/management.go

```diff
@@ -102,81 +112,168 @@ func NewManagement(parentCtx context.Context, config Config, provider Provider,
+	}
+    // private management
+	committeeTerm := getCommittee(systemRef, data.Committees)
+	subTerm := getSubscriptionStatus(systemRef, data.Subscriptions)
+	pvTerm := getProtocolVersion(systemRef, data.ProtocolVersions)
+	return maxOf(committeeTerm.AsOfReference, subTerm.AsOfReference, pvTerm.AsOfReference)
+}
+
+func maxOf(a, b, c primitives.TimestampSeconds) primitives.TimestampSeconds {
+	if a > b {
+		if a > c {
+			return a
+		} else {
+			return c
+		}
+	} else if b > c {
+		return b
+	} else {
+		return c
+	}
+}
+
 func (s *service) GetCurrentReference(ctx context.Context, input *services.GetCurrentReferenceInput) (*services.GetCurrentReferenceOutput, error) {
 	s.RLock()
 	defer s.RUnlock()
 	return &services.GetCurrentReferenceOutput{
-		CurrentReference: s.data.CurrentReference,
+		CurrentReference: getCurrentReference(input.SystemTime, s.data),
 	}, nil
 }
 
 func (s *service) GetGenesisReference(ctx context.Context, input *services.GetGenesisReferenceInput) (*services.GetGenesisReferenceOutput, error) {
 	s.RLock()
 	defer s.RUnlock()
 	return &services.GetGenesisReferenceOutput{
-		CurrentReference: s.data.CurrentReference,
+		CurrentReference: getCurrentReference(input.SystemTime, s.data),
 		GenesisReference: s.data.GenesisReference,
 	}, nil
 }
 
-func (s *service) getProtocolVersion(refTime primitives.TimestampSeconds) *ProtocolVersionTerm {
+/*
+ * Event data helper functions
+ */
+func (s *service) tryGetCurrentData(referenceTime primitives.TimestampSeconds) (*VirtualChainManagementData, error) {
 	s.RLock()
 	defer s.RUnlock()
-	i := len(s.data.ProtocolVersions) - 1
-	for ; i > 0 && refTime < s.data.ProtocolVersions[i].AsOfReference; i-- {
+	if referenceTime > s.data.EndPageReference {
+		return nil, errors.Errorf("ReferenceTime %d is in the future (current last valid is %d)", referenceTime, s.data.EndPageReference)
+	}
+	if referenceTime >= s.data.StartPageReference {
+		return s.data, nil
 	}
-	return &s.data.ProtocolVersions[i]
+	return nil, nil
 }
 
-func (s *service) GetProtocolVersion(ctx context.Context, input *services.GetProtocolVersionInput) (*services.GetProtocolVersionOutput, error) {
-	return &services.GetProtocolVersionOutput{
-		ProtocolVersion: s.getProtocolVersion(input.Reference).Version,
-	}, nil
+func (s *service) tryGetHistoricData(ctx context.Context, referenceTime primitives.TimestampSeconds) (*VirtualChainManagementData, error) {
+	s.Lock()
+	defer s.Unlock()
+	if referenceTime < s.cachedHistoricData.StartPageReference || referenceTime > s.cachedHistoricData.EndPageReference {
+		historic, err := s.provider.Get(ctx, referenceTime)
+		if err != nil {
+			return nil, errors.Wrapf(err, "management provider failed to get historic reference data for reference-time %d", referenceTime)
+		}
+		s.cachedHistoricData = historic
+		s.metrics.pageCachedStartRefTime.Update(int64(s.cachedHistoricData.StartPageReference))
+		s.metrics.pageCachedEndRefTime.Update(int64(s.cachedHistoricData.EndPageReference))
```

Comment:
do we want to count cache hits vs misses?

Reply:
efficiency is for next version. the metrics page is already very full of stuff